pipeline.py

```diff
@@ -89,8 +91,10 @@ def __init__(self):
 
     def process(self, item):
         global UAX, PFG
+        UAX_PFG_LOCK.acquire()
         if self._counter > 0:
             self._counter -= 1
+            UAX_PFG_LOCK.release()
```

Comment:
this would need to present before every return to prevent a deadlock
fortunately you can just put the entire function body into a `with UAX_PFG_LOCK:` block